### ISSUE TITLE
Memory leak with rapidly changing URL

### ISSUE THREAD

**Issue by yshui** [Drive4ik/simple-tab-groups#371]
## Steps to reproduce

 1. Start Firefox with clean profile.
 3. Install Simple Tab Group
 5. Goto this page: http://matthewrayfield.com/articles/animating-urls-with-javascript-and-emojis/
 6. Click on one of the check boxes to enable URL changing javascript

## Actual result

WebExtensions process consume a lot more CPU compared to when the plugins are disabled. Memory consumption firefox process increases rapidly.

## Environment

 * Platform (OS): Linux
 * Version of Firefox: 66.0.3
 * Version of STG: 3.4.4

**Comment by Drive4ik** [comment]
This happens because every 50ms the tab URL is changed and an event is raised that saves the tabs every 500ms to memory in order to save the last url of the tab. Tell me, why do this?
By the way, I did not notice the increased memory usage.

**Comment by yshui** [comment]
> Tell me, why do this?
I didn't create that monstrosity :)

So, is STG never going to be efficient against this kind of web page behavior? The memory usage problem didn't happen for you, but over here it causes firefox to use up all my memory. Feels like something a malicious site might make use of.

**Comment by Drive4ik** [comment]
The fact is that I don’t know how to distinguish a very frequent URL change for useful or harmful purposes... There are too many conditions for false positives and incorrect work of the addon...

**Comment by yshui** [comment]
Is it possible to not save tabs every 500ms? For example, could STG save the urls/tabs only when it's required?

**Comment by Drive4ik** [comment]
It works like this now; it does not save every 500ms, but saves data 500ms after the last tab changes. For example, the url of the tab 1 was updated, another tab was loaded after 450ms, and then another 300ms was updated, the url of another tab was updated, and only after this last change, plus another 500ms wait, the data is saved.

**Comment by yshui** [comment]
If that's the case, the CPU usage probably shouldn't be that high? Is this change included in 3.4.4?

**Comment by Drive4ik** [comment]
Yes, and for a long time it is
https://github.com/Drive4ik/simple-tab-groups/blob/056758b6c8bbb765faa384aaa03edb34bf0d035b/addon/src/background.js#L35-L52

**Comment by yshui** [comment]
So in the onUpdated handler 
https://github.com/Drive4ik/simple-tab-groups/blob/faa23d99a1c65a9133e5c31cf356eacaf633ea8c/addon/src/background.js#L1090-L1110

A 'group-updated' message is sent every time, could that be the reason?

**Comment by Drive4ik** [comment]
Maybe if you keep the add-on popup open, or the group management page...

**Comment by Drive4ik** [comment]
Try to test it like this:
![Screenshot_2](https://user-images.githubusercontent.com/7843031/56357565-c3631880-61e4-11e9-986f-a85ffc07cd2f.png)

**Comment by yshui** [comment]
I just found out something _really_ weird.

I just tried to reproduce this problem again, but it didn't happen. So I poked around a bit, and I found out that:

1. Restart firefox, enable just STG. No problem with changing URLs
2. From step 1, also enable TreeStyleTab. Now I can reproduce this problem.
3. Now disable TST, but keep STG enabled. This problem _still_ happens.
4. Disable STG, and this problem is gone
5. Enable STG, keep TST disabled, this problem is back.

And as for the numbers in `about:performance`, the "Energy Impact" field becomes "High" when this problem happens, the number is ~40. The memory usage shown doesn't increase, but in my system's resource monitor (`top`), I can clearly see the memory usage of "WebExtensions" and "firefox" increasing.
